eth_abi/utils/grammar.py

```diff
@@ -0,0 +1,325 @@
+import functools
+import re
+
+from parsimonious import expressions
+import parsimonious
+
+from eth_abi.exceptions import (
+    ABITypeError,
+    ParseError,
+)
+
+
+grammar = parsimonious.Grammar(r"""
+type = tuple_type / basic_type
+
+tuple_type = "(" type next_type* ")"
+next_type = "," type
+
+basic_type = base sub? arrlist?
+
+base = alphas
+
+sub = two_size / digits
+two_size = (digits "x" digits)
+
+arrlist = (const_arr / dynam_arr)+
+const_arr = "[" digits "]"
+dynam_arr = "[]"
+
+alphas = ~"[a-z]+"
+digits = ~"[0-9]+"
+""")
+
+
+class NodeVisitor(parsimonious.NodeVisitor):
+    """
+    Parsimonious node visitor which performs both parsing of type strings and
+    post-processing of parse trees.  Parsing operations are cached.
+    """
+    grammar = grammar
+
+    def visit_tuple_type(self, node, visited_children):
+        # Ignore left and right parens
+        _, first, rest, _ = visited_children
+
+        return TupleType((first,) + rest, node=node)
+
+    def visit_next_type(self, node, visited_children):
+        # Ignore comma
+        _, abi_type = visited_children
+
+        return abi_type
+
+    def visit_basic_type(self, node, visited_children):
+        base, sub, arrlist = visited_children
+
+        return BasicType(base, sub, arrlist, node=node)
+
+    def visit_two_size(self, node, visited_children):
+        # Ignore "x"
+        first, _, second = visited_children
+
+        return first, second
+
+    def visit_const_arr(self, node, visited_children):
+        # Ignore left and right brackets
+        _, int_value, _ = visited_children
+
+        return (int_value,)
+
+    def visit_dynam_arr(self, node, visited_children):
+        return tuple()
+
+    def visit_alphas(self, node, visited_children):
+        return node.text
+
+    def visit_digits(self, node, visited_children):
+        return int(node.text)
+
+    def generic_visit(self, node, visited_children):
+        if isinstance(node.expr, expressions.OneOf):
+            # Unwrap value chosen from alternatives
+            return visited_children[0]
+
+        if isinstance(node.expr, expressions.Optional):
+            # Unwrap optional value or return `None`
+            if len(visited_children) != 0:
+                return visited_children[0]
+
+            return None
+
+        return tuple(visited_children)
+
+    @functools.lru_cache(maxsize=None)
```

Comment:
Any reason for the `None`?  Seems like it should be fine to put an arbitrary upper bound on this (1024?) so that we **know** it can't grow indefinitely.

Reply:
I just figured that, in most systems using eth-abi, there would probably on the order of a few hundred type strings which would ever be parsed.  Also, with `maxsize=None`, `functools.lru_cache` foregoes a lot of machinery that it would normally use to properly evict things from the cache.  Since I figured the cache would probably never grow too large, I thought it would be better to opt for the speedier version which doesn't limit size.